### ISSUE TITLE
[GR 13] Integrate DID:Element ("Sidetree for Ethereum") into BRI-1

### ISSUE THREAD

**Issue by bumblefudge** [eea-oasis/baseline-grants#38]
## Grant name
Element-BRI-1 Integration

## Applicant background
[Spruce Systems](https://spruceid.com/) (NYC) is a cross-chain decentralized identity studio with deep roots in the Ethereum community, founded by Wayne Chang (@wyc) and Gregory Rocco (@obstropolos).  Our work is driven by standards-conformant technologies that build in agility and forward-compatibility by taking the most interoperable and disciplined interpretation of emerging standard.  Our focus on web3 and complex integrations across unconventional topographies makes Baseline very interesting for us as a bridge between these two sectors of the blockchain space. We are passionate about bringing verifiable credentials and decentralized addressing capabilities to production systems far and wide, and Baselining those systems seems a great opening to do so.

## Motivation and overview
As explained in #30 , Spruce Systems is putting forward a proposal to integrate into BRI-1 the results of GR12 (by Transmute Industries, with whom we've been coordinating). This will give BRI-1 DID capabilities right "out of the box" that are not only conformant with the recently-ratified DID core specification at W3C, but also V1 of the Sidetree Protocol. This would expand the possible architectures and configurations available to Baseline, for the negligible complexity and compute cost of another few docker containers.

## Execution risks
Our main concern is aspects of the existing BRI-1 NATS integration and phonebook design we have not yet understood-- but we were quite encouraged on today's call to hear Kyle express interest, as his input will be invaluable if either system presents challenges not visible from a distance.

## Downsides
The biggest downside to moving from UUIDs to DIDs is too much extensibility and optionality, which can present footguns for the developer with a dangerous combination of curiosity and no DID experience.  For this reason, we are proposing a "full autofocus" configuration, which we will demo in our hello-world tutorial and video; full support for did:web, did:eth, and arbitrary additional did methods will be explained in the documentation but *not* enabled by default.

## Tasks and schedule (milestones if applicable)
See slides 4 and 5 in the [pitch deck](https://github.com/eea-oasis/baseline-grants/files/6876877/26July.-.Baseline.Pitch.Deck.pdf) we prepared for today's meeting.  We are proposing to coincide the entirety of our integration work with the "Milestone 2" period of Transmute's work on GR12 (i.e. Q4), such that iteration and refinement on both sides can benefit from additional eyes and perspectives.  We can thus spend Milestone 1 analyzing the phonebook and NATS issues without rushing any collaborators on the Baseline side.

## Budget and justification
U$20K - It will take considerable engineering effort to ensure our branch is working as designed and without consequences for other systems in the BRI-1 implementation, and documentation appropriate to this context.

**Comment by Consianimis** [comment]
Coul we clarify the choice of integration into Provide's BRI-1 instead of Baseline Protocol CORE v0.1 or Consensys' BRI-2 ?
If the reason is that CORE V.1 is still in development and has not been released yet, how easy would it be to integrate DID:Element ("Sidetree for Ethereum") to CORE V.1 to the benefit of the wider community?

**Comment by bumblefudge** [comment]
We were invited by Andreas to do a quick assessment of BRI-1 and BRI-2 and figured integrating into both on a tight schedule for 20K would be an unrealistic ask.  We are more familiar with Nchain than the Codefi Orchestrate code, and we found the KMS aspects easier to get our hands around, so we felt more confident committing to doing the work in the time and budget constraints for just BRI-1.  One slide 4, we break down the components we've already reviewed and how we would modify them; making the equivalent changes to the BRI-1 equivalents of each component is presumably not that much more work, it was just harder to quote for running code + thorough documentation without further assesment.

Furthermore, nothing I see in the Core V.1 specification looks unsurmountable for extending the approach, but on these terms it's easier to feel confident about a branch on stable code.  Perhaps if the Core v.1 working group were willing to invite us to a meeting and do a check-in midway through the grant period and advise on whether our approach is too specific to BR-1 components, we could address any issues or assumptions in our documentation?  It seems like answering the question of how much work it will be to reconcile the BRI-1 prototype with Core v.1 specifications/abstractions would be easier to answer in detail after the grant, and a driving motivation for it!
